### PR TITLE
Show skills according to the group chosen by the user (Fixed #1539)

### LINKED ISSUE #1539: Add feature to load skills of a specific group when a group name is tapped
[Parent issue #1467 ]

**Actual Behaviour**

Skills listing UI flow has been changed to show skills based on metrics and the groups at the bottom of this page. The user cannot see the skills for a particular group after #1518  and #1522 . 

**Expected Behaviour**

Load all the skills belonging to a particular group (from the groups list provided at the bottom of the page), when it is tapped.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by arundhati24**
Fixes #1539  [Parent issue #1467]

Changes: The skills belonging to a particular group are loaded in a new fragment when the user taps that group.

Screenshots for the change: 

screen shot 2018-07-17 at 2 43 45 pm

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -85,11 +85,20 @@
...
         <activity
             android:name=".chat.YouTubeActivity"
-            android:theme="@style/Theme.AppCompat.Dialog">
-        </activity>
+            android:theme="@style/Theme.AppCompat.Dialog" />
+
         <meta-data
             android:name="com.google.android.youtube.API_KEY"
-            android:value="${YOUTUBE_API_KEY}"/>
+            android:value="${YOUTUBE_API_KEY}" />
```
Comment (batbrain7) [inline app/src/main/AndroidManifest.xml]: unnecessary change

File: app/src/main/java/org/fossasia/susi/ai/data/GroupWiseSkillsModel.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        val queryObject = SkillsListQuery(group, language, "true", "descending", "rating")
+        authResponseCallSkills = ClientBuilder.fetchListSkillsCall(queryObject)
+
+        authResponseCallSkills.enqueue(object : Callback<ListSkillsResponse> {
+            override fun onResponse(call: Call<ListSkillsResponse>, response: Response<ListSkillsResponse>) {
+                listener.onSkillFetchSuccess(response, group)
+            }
+
+            override fun onFailure(call: Call<ListSkillsResponse>, t: Throwable) {
+                t.printStackTrace()
```
Comment (batbrain7) [inline app/src/main/java/org/fossasia/susi/ai/data/GroupWiseSkillsModel.kt]: Timber

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsActivity.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.fossasia.susi.ai.skills.SkillsActivity
+
+/**
+ *
+ * Created by arundhati24 on 16/07/2018.
+ */
+class GroupWiseSkillsActivity : AppCompatActivity() {
+
+    private val TAG_GROUP_WISE_SKILLS_FRAGMENT = "GroupWiseSkillsFragment"
+
```
Comment (batbrain7) [inline app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsActivity.kt]: extra line

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (arundhati24) [reply app/src/main/AndroidManifest.xml]: This is not an unnecessary change, I added an activity and as usual used default refactoring by pressing cmd+option+L in Android Studio. This should have been done in the PR where this change was made, and I think there is no point in sending a separate PR for such spacing/style issues in a line or two.

**Comment by arundhati24** [comment]
@iamareebjamal Would you please guide me on the first review by @batbrain7 , regarding the GroupWiseSkillsActivity? Are the changes made by me fine or is there an issue with creation of a new activity? It would be very helpful and I would be able to make the necessary changes then.

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (iamareebjamal) [reply app/src/main/AndroidManifest.xml]: Not true, even though it should have been sent in original PR, stylistic changes not related to PR pollute review area and git blame and commit history

**Comment by iamareebjamal** [comment]
Also, yes, fragments should be preferred over activity

**Comment by arundhati24** [comment]
Okay @iamareebjamal I would make the change and update the PR. Thanks :)

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (arundhati24) [reply app/src/main/AndroidManifest.xml]: @iamareebjamal I would remove this change here then and fix it as part of the refactoring issue already opened.

**Comment by batbrain7** [comment]
@arundhati24 if you are adding a new feature and the code is not yet merged then why should an improper code be merged and then refactored again ?

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (batbrain7) [reply app/src/main/AndroidManifest.xml]: @arundhati24 if a new feature is to be added then it should be added properly rather than refactor it later as it is not dependent on the server neither the hardware so this will be fixed for a long time .

Comment (arundhati24) [reply app/src/main/AndroidManifest.xml]: @batbrain7 Please read carefully. I just said I will send a new PR fixing these spacing issues as part of refactoring issue already opened by me. And, this I have mentioned for the "unnecessary change". I will be sending a new PR for all spacing issues in some time.

Comment (arundhati24) [reply app/src/main/AndroidManifest.xml]: And, I would refactor all the code changes that I have made in this PR, as usual.

Comment (batbrain7) [reply app/src/main/AndroidManifest.xml]: @arundhati24 yes I deleted the comment immediately if you did not check

Comment (arundhati24) [reply app/src/main/AndroidManifest.xml]: We do get a mail. Please check things before commenting.

Comment (batbrain7) [reply app/src/main/AndroidManifest.xml]: I am sorry , I will see to it from now

**Comment by arundhati24** [comment]
@iamareebjamal @batbrain7 I removed the GroupWiseSkillsActivity and attached the GroupWiseSkillsFragment to the Skills Activity and also made other changes requested by @batbrain7. Please have a look at it.

**Comment by arundhati24** [comment]
@iamareebjamal Would you please review this PR, so that the necessary changes, if any, can be made and the next release can be made? It would be very helpful.

File: app/src/main/java/org/fossasia/susi/ai/dataclasses/GroupWiseSkills.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.fossasia.susi.ai.rest.responses.susi.SkillData
+
+/**
+ *
+ * Created by arundhati24 on 16/07/2018.
+ */
+data class GroupWiseSkills(
+        var group: String,
+        var skillsList: MutableList<SkillData>
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/dataclasses/GroupWiseSkills.kt]: Why are these var, not val

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsFragment.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        mLayoutManager.orientation = LinearLayoutManager.VERTICAL
+        groupWiseSkills.layoutManager = mLayoutManager
+        skillsAdapter = SkillsListAdapter(requireContext(), skills)
+        groupWiseSkills.adapter = skillsAdapter
+        groupWiseSkills.onFlingListener = null
+        skillAdapterSnapHelper.attachToRecyclerView(groupWiseSkills)
+    }
+
+    override fun visibilityProgressBar(boolean: Boolean) {
+        if (boolean) progressSkillWait.visibility = View.VISIBLE else progressSkillWait.visibility = View.GONE
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsFragment.kt]: `progressSkillWait.visibility = if (show) View.VISIBLE else View.GONE`

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    override fun onSkillFetchSuccess(response: Response<ListSkillsResponse>, group: String) {
+        groupWiseSkillsView?.visibilityProgressBar(false)
+        if (response.isSuccessful && response.body() != null) {
+            Timber.d("GROUP WISE SKILLS FETCHED")
+            val responseSkillList = response.body().filteredSkillsData
+            if (responseSkillList != null && responseSkillList.isNotEmpty()) {
+                skills.skillsList.clear()
+                skills.skillsList = responseSkillList as MutableList<SkillData>
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsPresenter.kt]: Why the need of cast?

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/adapters/recycleradapters/SkillsListAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+ * Created by arundhati24 on 16/07/2018.
+ */
+class SkillsListAdapter(val context: Context, private val skillDetails: GroupWiseSkills) : RecyclerView.Adapter<SkillViewHolder>() {
+
+    private val imageLink = "https://raw.githubusercontent.com/fossasia/susi_skill_data/master/models/general/"
+
+    @NonNull
+    override fun onBindViewHolder(holder: SkillViewHolder, position: Int) {
+        if (skillDetails != null) {
+            val skillData = skillDetails.skillsList.toTypedArray()[position]
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/adapters/recycleradapters/SkillsListAdapter.kt]: Why not `get`

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/adapters/recycleradapters/SkillsListAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                holder.skillExample?.text = StringBuilder("\"").append("\"")
+            else
+                holder.skillExample?.text = StringBuilder("\"").append(skillData.examples[0]).append("\"")
+
+            if (skillData.image == null || skillData.image.isEmpty()) {
+                holder.skillImage?.setImageResource(R.drawable.ic_susi)
+            } else {
+                Picasso.with(context.applicationContext).load(StringBuilder(imageLink)
+                        .append(skillDetails.skillsList[position].group.replace(" ", "%20"))
+                        .append("/en/").append(skillData.image).toString())
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/adapters/recycleradapters/SkillsListAdapter.kt]: Form the URL above in separate variable and use it here

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsPresenter.kt (hunk 6 above)
Comment (arundhati24) [reply app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsPresenter.kt]: @iamareebjamal Because the list received in the response object is not mutable. But, the list 'skillsList' in the 'skills' object is defined as a mutable list. I had to define this as a mutable list because I need to add two more strings (Categories and All) at the top of the list.

File: app/src/main/java/org/fossasia/susi/ai/dataclasses/GroupWiseSkills.kt (hunk 4 above)
Comment (arundhati24) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/GroupWiseSkills.kt]: Because these are being assigned values when we get response from the server.
